packages/diagram/lib/diagram.dart

```diff
@@ -0,0 +1,442 @@
+    assert(duration != null);
+    assert(duration >= Duration.zero);
+    _timestamp += duration;
+
+    handleBeginFrame(_timestamp);
+    handleDrawFrame();
+  }
+}
+
+/// A generator used by [DiagramController] for generating filenames for frames
+/// in animated diagram sequences.
+typedef File AnimationFrameFilenameGenerator(Duration timestamp, int index);
+
+/// A controller for creating diagrams generated by using Flutter widgets.
+///
+/// This is used to configure and create individual image diagrams, as well as
+/// animations of diagrams with control over the time used by animation tickers.
+class DiagramController {
+  /// Creates a diagram controller.
+  ///
+  /// The [builder] parameter must not be null, and is required.
+  DiagramController({
+    @required WidgetBuilder builder,
+    this.outputDirectory,
+    AnimationFrameFilenameGenerator frameFilenameGenerator,
+    double pixelRatio: 1.0,
+    Size screenDimensions: _kDefaultDiagramViewportSize,
+  }) : assert(builder != null) {
+    this.frameFilenameGenerator = frameFilenameGenerator;
+    outputDirectory ??= Directory.current;
+    _binding.pixelRatio = pixelRatio;
+    _binding.screenDimensions = screenDimensions;
+    this.builder = builder;
+  }
+
+  /// The builder used to generate each frame of the diagram.
+  ///
+  /// If the [builder] is changed, then the widget it builds will be used for
+  /// subsequent frames drawn of the diagram.
+  ///
+  /// Defaults to the `initialBuilder` given to [new DiagramController].
+  WidgetBuilder get builder => _builder;
+  WidgetBuilder _builder;
+  set builder(WidgetBuilder builder) {
+    assert(builder != null);
+    if (_builder != builder) {
+      _builder = builder;
+      _binding.updateDiagram(builder);
+    }
+  }
+
+  /// The output directory used when a [File] generated by the
+  /// [frameFilenameGenerator], or the `outputFile` argument to
+  /// [drawDiagramToFile] is a relative path. The written file will be relative
+  /// to [outputDirectory]. If the filename is absolute, then [outputDirectory]
+  /// is ignored.
+  ///
+  /// Defaults to [Directory.current].
+  ///
+  /// On a device, this will need to be set to the result of
+  /// [getApplicationDocumentsDirectory] (from the path_provider package) or a
+  /// subdirectory of that directory to have sufficient permissions to write
+  /// files.
+  Directory outputDirectory;
+
+  /// The generator for filenames when calling drawAnimatedDiagramToFile.
+  ///
+  /// If the returned filenames are relative paths, they will be relative to
+  /// [outputDirectory].
+  AnimationFrameFilenameGenerator get frameFilenameGenerator => _frameFilenameGenerator;
+  AnimationFrameFilenameGenerator _frameFilenameGenerator = _basicFrameFilenameGenerator;
+  set frameFilenameGenerator(AnimationFrameFilenameGenerator frameFilenameGenerator) {
+    _frameFilenameGenerator = frameFilenameGenerator ?? _basicFrameFilenameGenerator;
+  }
```

Comment:
this setter violates the rule that `a = b` should mean that `a == b` for all values of b, when `b == null`.